Feat/divset pmtiles

awdem opened the pull request:
This PR adds a pipeline for creating pmtiles for divisionsets and exposes them on the frontend.

It adds a few new dependencies so you'll need to run `uv sync`, `npm install`, and install [tippecanoe](https://github.com/felt/tippecanoe?tab=readme-ov-file#installation).

_(Note: I've only added tippecanoe to CI at the moment, so that it can run the new tests I wrote. So this PR isn't ready to be merged as is, but after I've gotten feedback, I'll set up deployment properly.)_

The PR adds a command `create_pmtiles_for_divset` that uses a `PMtilesCreator` class to create a pmtiles file for a given divisionset id. The pmtiles file will be saved locally in `STATIC_ROOT/pmtiles-store` or to the S3 bucket at the `PUBLIC_DATA_BUCKET ` constant if it's set in settings.

It also adds a PMtilesView that serves the pmtiles from either static or S3 and a new divisionset detail view with a maplibre-gl map that displays them.

The org page now just lists divisionsets as links to detail pages:
<img width="924" height="942" alt="image" src="https://github.com/user-attachments/assets/cf269023-641c-4b72-a155-e4b9b864fae9" />

The new divisionset detail page with the map element folded:
<img width="859" height="826" alt="image" src="https://github.com/user-attachments/assets/be55a7a5-cac3-46e8-80e4-ca305f204069" />
And unfolded:
<img width="565" height="877" alt="image" src="https://github.com/user-attachments/assets/8cbfe497-3a0f-4a4d-b3fb-57971d0f692a" />
Clicking on a division in the map will display its name.
![mapgif](https://github.com/user-attachments/assets/fd4da5f2-fee0-40ca-b4cb-5899971db7b1)


## Testing

The command works out of the box locally, but if you want to test writing and reading from S3, you'll need to set PUBLIC_DATA_BUCKET ="ee.public.data" in your `.env` file and use the AWS credentials for the production EE account.


## Follow-up Work for this PR:

- Add tippecanoe to CI cache
- Add tippecanoe to the app dependencies for deployment
- Make sure dev and stage have a PUBLIC_DATA_BUCKET on S3
- Set PUBLIC_DATA_BUCKET on AWS Parameter store in all envs
- Once merged, run create_pmtiles_for_all_divsets --overwrite on prod
- Add a pmtiles creation step to the ECO processing wiki.


## Future Work:
- Currently we're still using leaflets-js to display org and election maps. I'd like to remove leaflets-js in favor of maplibre. 
- I'd also like to add another map to the frontend on which you can overlay all the divisionsets for the same organisation, with toggles for each divisionset. This is only really for me to use when importing boundaries so it only needs to show up if DEBUG=True.
- Implement pmtiles creation for divsets with >1 division_subtypes (sp, old senedd, gla?)
- Building PMTiles in response to changes (and/or on a schedule)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210054691908446

awdem commented on every_election/apps/organisations/models/divisions.py:
```diff
@@ -38,6 +38,14 @@ def __str__(self):
+
+    @property
+    def pmtiles_s3_key(self):
+        return f"pmtiles-store/{self.pmtiles_file_name}"
```
I've hardcoded the folder on s3 to always be pmtiles-store, but I'm not sure if that should also be a constant since it would hypothetically be annoying to change later?

chris48s reviewed (commented):
This is really good.

I do have quite a lot of comments. This is mainly a symptom of the fact that there's a lot of work here, rather than because what's here is bad.

chris48s commented:
There are a few things where I'm not really sure where to leave the comment on the diff:

1. The styling here is a bit ropey:

<img width="178" height="72" alt="Screenshot at 2025-08-20 11-04-14" src="https://github.com/user-attachments/assets/0452a9d4-e154-4a9c-93e5-32f81f42493f" />

2. Given we may end up in a situation where a divisionset exists but we can't render a map because the tiles haven't been generated, can we handle that gracefully rather than just rendering a big empty gap.

3. Could we include other things in the popup as well as the name (e.g: GSS code if we have one)?

Finally, there is one other thing I want to add to the follow up work (not for this PR):

- Building PMTiles in response to changes (and/or on a schedule)

awdem commented on every_election/apps/organisations/views/public.py:
```diff
@@ -100,3 +105,34 @@ class SingleBoundaryReviewView(DetailView):
+        if settings.PMTILES_BUCKET:
+            return HttpResponseRedirect(
+                f"https://s3.eu-west-2.amazonaws.com/{settings.PMTILES_BUCKET}/{divset.pmtiles_s3_key}"
+            )
```
That makes sense. Done in [38b3eee](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/38b3eee91b3bbb67aafed36c1d5650c711c5341d)

awdem commented on every_election/apps/organisations/templates/organisations/divisionset_map.html:
```diff
@@ -0,0 +1,111 @@
+                        "id": "divset-line",
+                        "type": "line",
+                        "source": "divset",
+                        "source-layer": "{{ divisionset.id }}",
```
I did think it looked a bit weird to mix the two up like that. I couldn't find a `|tojson` template filter but I did find `|json_script` in the docs, so I used the [pattern here](https://docs.djangoproject.com/en/5.2/ref/templates/builtins/#json-script) to implement something like the above in [8ae9161](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/8ae9161564322dc8e59700a22a9f88ce6a85a257).

I think it's the same thing?

chris48s replied:
Great. Yes sorry `|tojson` is for jinja templates, not django :facepalm: 
In django you can either do it with json_script or [escapejs](https://docs.djangoproject.com/en/5.2/ref/templates/builtins/#escapejs)

What you've done here is perfect :+1:

awdem commented on every_election/apps/organisations/pmtiles_creator.py:
```diff
@@ -0,0 +1,64 @@
+            # TODO: handle this error better
+            print(f"Error creating pmtiles file: {e}")
+
+    def extract_geojson(self, dest_dir):
```
Done in d0fe3d18f3a6314c1f9e8058071598979c41763a

awdem commented on every_election/apps/organisations/management/commands/create_pmtiles_for_divset.py:
```diff
@@ -0,0 +1,97 @@
+                    f"OrganisationDivisionSet with id '{divset_id}' does not exist."
+                )
+            )
+            return
```
I've split these changes over two fixup commits:

[ef66acd](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/ef66acd8a68ebd0aae12c32098836609870e502a) refactors `create_pmtiles_for_divset` to raise a CommandError if divisionset validation fails rather than just logging an error and returning 0.

[03866f1](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/03866f1c303ab8ce16c978339a764ba607b8a77f) implements your suggestion for handling individual failures of the  `create_pmtiles_for_divset` command for `create_pmtiles_for_all_divsets`

awdem commented on every_election/apps/organisations/templates/organisations/organisation_detail.html:
```diff
@@ -7,86 +7,46 @@
 
             {% include "organisations/organisation_summary.html" %}
+            {% if object.divisionset.exists %}
+                <h2>Sub-divisions</h2>
```
Yeah I feel like "Division Sets" might not mean much the general public; but, on the other hand, if you're already looking at an organisation page on purpose, then you're more likely to understand it. I also can't think of anything better.  

Changed Sub-divisions -> Division Sets in 1e0d091bcc981392cb19f15e594e9498fe2a6e69

chris48s replied:
If we think of a better word, we can change it later

awdem commented on every_election/apps/organisations/tests/test_pmtiles_view.py:
```diff
@@ -0,0 +1,52 @@
+PMTILES_BUCKET = "test-pmtiles-store"
+
+
+class TestPMTilesView(TestCase):
```
Added a test in https://github.com/DemocracyClub/EveryElection/pull/2498/commits/7abffc54e668c97cc150bf672e78ab4a5fa68430 along with a small refactor to DRY things up.

chris48s replied:
lovely :+1:

awdem commented on every_election/apps/organisations/management/commands/create_pmtiles_for_divset.py:
```diff
@@ -0,0 +1,97 @@
+from organisations.pmtiles_creator import PMtilesCreator
+
+
+# TODO: implement optional overwrite arg?
```
I was on the fence, but I ended up doing it in [f4df764](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/f4df7647cc3b14b81cfdb7e40f362bee3e5c2321) .

awdem commented on every_election/apps/organisations/templates/organisations/divisionset_map.html:
```diff
@@ -0,0 +1,111 @@
+                    zoom: h.maxZoom - 2,
+                    center: [h.centerLon, h.centerLat],
+                    style: style,
+                });
```
Ok, so after trying to mess with the pmtiles generation for a while, I've basically copied what Sym did in the [EONI maps](https://github.com/DemocracyClub/eoni-maps/blob/main/web/src/lib/MapViewer.svelte#L188-L201) and fit the map bounds to the file bounds after the map loads. Changes here: https://github.com/DemocracyClub/EveryElection/pull/2498/commits/35207c66c6b1363038bc289de0e3dd50e6174060

awdem commented:
> 1. The styling here is a bit ropey:

This is being caused by a  CSS class between the design system and Maplibre. I'm sure there is a cleverer way to fix it but I just made it disappear with f688407c2ecc41adaf19fe27eb8ac0685db2d716.

> 2. Given we may end up in a situation where a divisionset exists but we can't render a map because the tiles haven't been generated, can we handle that gracefully rather than just rendering a big empty gap.

In order to implement this I decided to add another property to the DivisionSet model in 6d4e44e1fdc4dceea67e5abe96a26a3857fa2dce and then checking that property before displaying the map in 25286fd88c8a58be73f91abd6f97008250942ddb. I wasn't 100% confident that opening an s3 client in a property was a good idea. Let me know if there's a better way to do this.

> 3. Could we include other things in the popup as well as the name (e.g: GSS code if we have one)?

Add official_identifiers in f6af3ce59ff402cab55aa3f288315762e9da9b9b. Not sure if anything else if worth including? For my purposes of comparing boundaries, I don't think so.

> Finally, there is one other thing I want to add to the follow up work (not for this PR):
> 
> * Building PMTiles in response to changes (and/or on a schedule)

Added to the "Future Work" list.

chris48s commented on every_election/apps/organisations/management/commands/create_pmtiles_for_divset.py:
```diff
@@ -0,0 +1,97 @@
+                )
+                return
+
+        pmtile_creator = PMtilesCreator(divset)
```
Agreed :+1: Lets handle it in a follow up PR

chris48s commented on every_election/apps/organisations/templates/organisations/divisionset_map.html:
```diff
@@ -0,0 +1,130 @@
+                //navigation controls
+                map.addControl(new maplibregl.NavigationControl(), 'top-right');
+
+                const button = document.querySelector(".maplibregl-ctrl-attrib-button")
+                button.style.fontSize = 0;
```
A less hacky way to hide this would be to put

```css
.maplibregl-ctrl-attrib-button::before {
    display: none;
}
```

in `styles.scss`

Slightly more generally, if we going to make more use of maplibregl it would be nice to fix this in the design system and make a nice component for it, like this https://democracyclub.github.io/design-system/components/map-leaflet/
but not now. Not today.

awdem replied:
Agreed about the component. I was imagining that being part of the future work of removing leaflet-js since that also includes removing its component.

changes made in 85ecd14a86e66d4f1b99a313d3ff392708820625

chris48s commented on every_election/apps/organisations/management/commands/create_pmtiles_for_all_divsets.py:
```diff
@@ -0,0 +1,27 @@
+from organisations.models import OrganisationDivisionSet
+
+
+class Command(BaseCommand):
```
If I call `python manage.py create_pmtiles_for_all_divsets` and I've already made tiles, it will say

```
$ python manage.py create_pmtiles_for_all_divsets
Processing DivisionSet: 649
norfolk-649.pmtiles already exists locally. Skipping (use --overwrite to force).
Processing DivisionSet: 787
essex-787.pmtiles already exists locally. Skipping (use --overwrite to force).
```

but if I call `python manage.py create_pmtiles_for_all_divsets --overwrite`

`create_pmtiles_for_all_divsets` does not know the `--overwrite` flag

awdem replied:
Added `--overwrite` flag to create_pmtiles_for_all_divsets  in 2112605f65dd3ccdb2681ce4699844306cdf0b80

chris48s commented on every_election/apps/organisations/models/divisions.py:
```diff
@@ -38,6 +41,36 @@ def __str__(self):
+    def has_pmtiles_file(self):
+        if settings.PUBLIC_DATA_BUCKET:
+            s3_client = boto3.client("s3")
+            # avoiding circular import
+            from organisations.boundaries.lgbce_review_helper import (
+                check_s3_obj_exists,
+            )
```
I feel like this is a sign that `check_s3_obj_exists` is in the wrong place. It isn't really intrinsically anything to do with LGBCE. They're about talking to a file storage system.

I would say that `check_s3_obj_exists` and `upload_obj_from_url` should move into the `storage` app. Ideally they'd be methods of the `S3Wrapper` class, but if that is a pain, lets just move the functions/tests as they are into that app and update the imports here and in the LGBCE code.

awdem replied:
This took a bit more work than I expected and I'm wondering if  `S3Wrapper` should be renamed `S3BucketWrapper` now? It could also probably use more tests now.

Moved functions to S3Wrapper in 0affd6e48fbc3ce72356c8a7067ead45998c03ce

Adapted pmtiles pipeline to use s3wrapper in 41b452996ab6e60bcce742adf077bf0b81f5122c

chris48s commented on every_election/apps/organisations/models/divisions.py:
```diff
@@ -38,6 +41,36 @@ def __str__(self):
             self.pk, self.short_title, self.active_period_text
         )
 
+    @property
+    def has_pmtiles_file(self):
```
> I wasn't 100% confident that opening an s3 client in a property was a good idea

I think that is fine, but lets make it a `@cached_property` so if we end up calling it multiple times in the lifetime of the object we only call out to s3 once.

awdem commented:
Ok so this is nearly done. I've added tippecanoe via a Makefile to CircleCI in https://github.com/DemocracyClub/EveryElection/pull/2498/commits/7e4c7d67cc7fa0f2835b1d647fa2ff7feaaa1c5b

I've used the same Makefile to install tippecanoe as part of deploy in https://github.com/DemocracyClub/EveryElection/pull/2498/commits/4a219ba9acf8509e9b12c03f64b5ffad24cb1979.

The `install_tippecanoe` shell script installs tippecanoe into the default `INSTALL_DIR` (/usr/local) and uses make clean to remove the tar file and source directory afterwards. It's currently deployed to dev if you want to give it a go.


I'm not 100% happy with the process of caching in CI. There are two issues:
1) We're currently caching two copies of the binaries. This because `$(INSTALL_DIR)/bin/tippecanoe` runs `make` which makes the binaries in the source directory and `make install`, which copies them to the passed `INSTALL_DIR` (defaults to `/usr/local`). I want the`make install` to run when used in deployment so that the binaries end up in `usr/local`, but I don't want it to run in CI because I have to cache the source directory anyway. Have two sets of binaries doesn't make a huge difference to the cache size, but it would be nice to get it to work properly. 
2) The Makefile is still trying to make the final target (`$(INSTALL_DIR)/bin/tippecanoe`) each time and I'm not sure why. Presumably the target tippecanoe binary file has a modified timestamp that's older than its dependency the source directory? I added `sudo touch $@` to try and update the binary file timestamp but it hasn't changed anything. This doesn't happen locally so I'm wondering if its a quirk of CI.  It isn't really a big deal because, the target step doesn't rebuild the binaries, just copys them from the cached source directory to the cached `INSTALL_DIR` which takes no time (see [CI].(https://app.circleci.com/pipelines/github/DemocracyClub/EveryElection/3503/workflows/e26b2414-96e0-4980-a387-6785077a5f2c/jobs/8436)) . 





I didn't realize that Django Pipeline doesn't minify static files when `DEBUG=True`, so both [f6a9b8e](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/f6a9b8eb96a40fbe9683c04196dc7d099de67987) and [0c7d564](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/0c7d564232f6a514676161411b2658d3dfef6d8d) are addresssing issues that I found when I was testing deployment.

[061c412](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/061c4126967a31ed627700c680e93daa6bf52363) was just a typo I noticed.

chris48s commented on deploy/Makefile:
Minor point, but I don't think `deploy/Makefile` is the ideal place for this. `deploy/Makefile` implies to me it is the makefile for the whole deploy. I reckon this should go in a tippecanoe subdir

awdem replied:
Done in https://github.com/DemocracyClub/EveryElection/pull/2498/commits/1155f3f98723b4316ce1c168742e64c29bf23b7b and https://github.com/DemocracyClub/EveryElection/pull/2498/commits/21615c0b1fa7cc5a259d26943bf84d6d2514377b

chris48s commented:
Once we're happy with the process for installing tippecanoe can we also add a note on it to https://github.com/DemocracyClub/EveryElection/blob/master/INSTALL.md somewhere

awdem commented:
As discussed, I broke up the rule in the makefile that was building and installing tippecanoe in  [547432a](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/547432a19a4e3fd7e040392fc511069eb44a66e8).

I also gave a phony `build_in_src_dir` rule so that I didn't have to add the actual path of the tippecanoe source directory into the CI config. This way, if we want to bump the tippecanoe version, we only have to change it in the Makefile. 

Finally rather than targeting the source directory for the rule that unzips the tar file (`$(INSTALL_DIR)/$(SRC_DIR)`), I'm instead targeting the Makefile inside of the source directory (`$(INSTALL_DIR)/$(SRC_DIR)/Makefile`). This avoids all the timestamp modification troubles I was having with CI. 

So I think I've addressed all the issues described previously.

awdem commented:
> Once we're happy with the process for installing tippecanoe can we also add a note on it to https://github.com/DemocracyClub/EveryElection/blob/master/INSTALL.md somewhere

Done in [330310d](https://github.com/DemocracyClub/EveryElection/pull/2498/commits/330310d244a480dae5299626304bacdc1a52877e)

chris48s reviewed (approved):
Nice. I reckon this is good to go :+1: In order to deploy this we are going to need to:

- Squash all the fixups to tidy up
- Make sure we've got a S3 bucket set up for dev/stage/prod
- Set the relevant env vars in each environment
- Deploy
- Run create_pmtiles_for_all_divsets after deploying

Can I leave it with you to co-ordinate all of that

..and then we've identified various bits of follow up work we would like to do
Can you also write up tickets for all those tasks
